### DISCUSSION TITLE
"headless" operation possible?

### DISCUSSION THREAD

**Discussion by dieterstoll** [hvianna/audioMotion.js#48] [General]
Hello, I'd like to be able to operate audioMotion (which, for the record, is one heck of a brilliant piece of software!) in a quasi-headless mode (not entirely, obviously - a display would be connected to the Pi 4, but not a keyboard or mouse). I am running the public/index.html file directly in a browser because all I need is microphone input (no sound files to play). While some of the preferences are apparently stored in a browser pref file, others (most prominently: microphone use) are not. I would like to ask if there is a way to override or modify the default settings so that when the  browser starts, the audioMotion.js file has got all the settings in place. Thanks in advance for any tips!

**Comment by hvianna** [comment]
Hello! I'm happy to know audioMotion is being useful to you!

You're correct, all settings are automatically saved and restored in the next session, except for the source and speakers settings.

I'm not sure whether you can allow persistent access to your microphone, so you don't have to accept the browser request every time, but I'll have a look into it this weekend. If it's possible, I'll add these to the auto-saved settings.

**Reply by dieterstoll** [reply]
¡Hola Henrique! Thanks so much for your quick reply, and for your readiness to deal with my somewhat exotic request. I am trying to create a somewhat artsy thing that will translate ambient sounds into a semi-scientific visual experience. It is supposed to be an embedded system without network access from the outside, so security is not really a concern here. Looking back on 45+ years of experience with computers and having written my first programs (including FFT and various filtering techniques) in Fortran on punched cards for a CDC 6600 mainframe back around 1976 I am pleased to see how those somewhat abstract number crunching techniques can be applied to create a playful yet scientifically accurate experience on hardware that fits in a shirt pocket. Keep up the superb work!

**Reply by hvianna** [reply]
Nice to meet another computing veteran! 😄 I started a bit later, around 1982, with an Apple ][ microcomputer and BASIC programming, and had the chance to work a little with an IBM mainframe during the 90s. It always blows my mind when I think about how much computing power has grown and devices have shrinked since then.

Well, I just made the necessary changes to save the Source and Speakers settings, but it seems you can't allow permanent access to the microphone when loading an html file directly - the browser will ask again for permission every time the file is loaded, and it also requires an extra click to unlock the audio context so the analyzer can receive audio data.

Everything seems to work fine if you're accessing via http though, so I think you'll need to run a simple web server on your Pi for this to work. See in the pictures below how the file version does not have access to the "Site settings" section, where you can set permissions.

| http (localhost)  | file  |
|---|---|
| ![image](https://github.com/hvianna/audioMotion.js/assets/1033735/18b52055-2a7b-4782-8e45-455743135993) | ![image](https://github.com/hvianna/audioMotion.js/assets/1033735/1f43d590-9a7c-4456-8293-336bf356a90d) |



Anyways, this feature will be included in the next version I expect to release in a couple of weeks, maybe earlier. I would love to see some pictures or a video of your experiment later!

Cheers!

**Reply by JesusGod-Pope666-Info** [reply]
It could be nice if one could setup the settings as to how the app starts out in settings at load up.
Like I have some preferred visualizing in mind, maybe even a standard startup playlist - where one could make a quick sound tutorial maybe.
Not something totally important and all, but it could be interesting to have the option to customize it for whatever options one wants out of the package.

**Reply by hvianna** [reply]
@JesusGod-Pope666-Info Yes, I plan to add something along those lines, via the new `config.json` file.

**Comment by hvianna** [comment]
@dieterstoll I initially forgot to add a mention to this in the changelog, but the feature is now available in [version 24.3](https://github.com/hvianna/audioMotion.js/releases/tag/v24.3)

**Reply by dieterstoll** [reply]
Thank you so much for picking up my suggestion so promptly! My faith in humanity is now partially restored :-)

**Reply by dieterstoll** [reply]
Looking quite good in nginx at this point - the only remaining item on the wish list is the Fullscreen switch. If the Fullscreen setting could also be saved...

**Reply by JesusGod-Pope666-Info** [reply]
Well, sadly more madness is going to come in the world. But it is nice to see there is still some sane people around.

**Reply by hvianna** [reply]
@dieterstoll I'm glad it's working! Unfortunately, fullscreen can only be triggered in response to a user action event, like a mouse click or key press. There's no way I can automate that.

The best we can do right now is using `"mediaPanel": "close"` in the [config.json file](https://audiomotion.app/docs/#/server?id=configjson-file) to start with an expanded analyzer area.

**Reply by dieterstoll** [reply]
Well... I found a solution for the Fullscreen problem, but it comes with a caveat attached. THIS WILL ONLY WORK if Wayland is disabled on the Pi (you can do this in raspi-config under Advanced Options). Another caveat: once Wayland has been disabled, the built-in VNC server won't work (it's Wayland specific), so disable that in raspi-config, too. If a VNC server is needed, install x11vnc or another one of your choice. With that out of the way, we can now use xdotool. 
`sudo apt install xdotool`
Once audioMotion runs inside chromium-browser:
`xdotool search --onlyvisible --name audioMotion windowfocus key f`
It is important to give the window focus before sending the simulated 'f' keypress. Chrome will only accept keyboard input if and when it has focus.

**Reply by hvianna** [reply]
That's awesome, thanks for sharing this solution!

**Comment by dieterstoll** [comment]
So... my project is nearly finished. Can't thank Henrique enough for his help. Had to jump through a lot of hoops to get everything to work under bookworm. Anyway I am quite satisfied now. Small unlisted video here: https://youtu.be/75wIm6cu9sU
I will gladly divulge details of the build if someone's interested.

**Reply by hvianna** [reply]
I'm glad you got it working, this looks so cool! 👏👏
If you'd like, feel free to make a post in the [Show & Tell](https://github.com/hvianna/audioMotion.js/discussions/categories/show-and-tell) section!
